build windows with conan support

nilsnolde opened the pull request:
fixes #4 

~~still WIP, didn't try on windows yet. worked on linux though.~~ windows worked as well as on linux. so anyone should be able to use conan if desired.

one thing I realized: at least on linux conan installs only static libraries for openssl & crypto, so the final python .so is bigger but not much to do for auditwheel then.

jonathf commented:
Cool.

Just so you are aware, `cibuildwheels` already does this. On linux it uses [auditwheel](https://github.com/pypa/auditwheel) and on macos it uses Delocate (I think).

Considering the short list of dependencies, I much prefer static linking.

jonathf commented:
If you change the branch trigger from `master` (which does not exist) to `nn-add-win-builds`, I belive I should get a prompt about allowing CI to kick in for your builds. When you think it is time.

nilsnolde commented:
the changes in the actions yaml are up for discussion for sure. indeed my muscle memory kicked in with master. but it's fine in the `pull_request` section, then it should kick in as well.

nilsnolde commented:
urgh.. it worked so far, but seems we're running into https://github.com/pybind/pybind11/issues/1908. running the example in ipython:
```
c:\users\nilsn\documents\dev\python\pyvroom\venv\lib\site-packages\vroom\input\input.py in set_durations_matrix(self, profile, matrix_input)
     88         assert isinstance(profile, str)
     89         if not isinstance(matrix_input, Matrix):
---> 90             matrix_input = Matrix(numpy.asarray(matrix_input, dtype="uint32"))
     91         _Input.set_durations_matrix(self, profile, matrix_input)

RuntimeError: Incompatible buffer format!
```

I didn't read the full issues, it's quite long, so there might be a solution/workaround somewhere. at least it worked locally.. pushing soon.

jonathf commented:
Ah, yeah I am familiar with the issue.

We need a smarter way to determine the dtype in `vroom.Matrix` initializers. Either by selecting it in a smarter way, or by doing some sort of platform check.

Shouldn't happen too often. Right now it should only be `vroom.Matrix` and `vroom.Amount` which I pushed Yesterday.

nilsnolde commented:
> Ah, yeah I am familiar with the issue.

ok great!

that should be it now I hope. I'll comment more in the PR. just one thing I don't quite get yet: how does cibuildwheel what's in `[build-system].requires` of the `pyproject.toml`? is there some magic going on inside cibuildwheel which uses that value? locally I had to install pybind etc in my virtual env (I have a `build-requirements.txt` for pip I thought I'd contribute). but then CI obviously works, just don't understand how;)

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -2,46 +2,46 @@ name: Wheels
       with:
         submodules: recursive
 
+    - name: Cache Conan dependencies
+      if: matrix.os == 'windows-latest'
+      uses: actions/cache@v2
+      with:
+        path: $USERPROFILE/.conan
+        key: conan-${{ hashFiles('conanfile.txt') }}
```
it's doing a source install which takes forever

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -2,46 +2,46 @@ name: Wheels
     - uses: pypa/cibuildwheel@v2.3.1
       env:
-        MACOSX_DEPLOYMENT_TARGET: 10.15
+        MACOSX_DEPLOYMENT_TARGET: 10.14
```
It seems 10.14. has full c++17 support already

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -2,46 +2,46 @@ name: Wheels
       env:
-        MACOSX_DEPLOYMENT_TARGET: 10.15
+        MACOSX_DEPLOYMENT_TARGET: 10.14
+        CIBW_TEST_COMMAND_WINDOWS: ''
```
I'll remove that again, the test is just the import..

nilsnolde commented on .gitignore:
```diff
@@ -8,7 +8,7 @@ __pycache__/
 # Distribution / packaging
 .Python
-build/
+*build/
```
conan installs to `conda_build`

nilsnolde commented on README.md:
```diff
@@ -19,14 +19,29 @@ VROOM can also solve any mix of the above problem types.
+## Building
+
+Building the source distributions on another OS requires:
+- the `./build-requirements.txt` Python dependencies
```
ah, didn't include that file yet, but would be helpful no?

nilsnolde replied:
I added it

nilsnolde replied:
I still think it can't hurt to have it here for the old people like me who are still used to `python setup.py xx`;) happy to remove it as well though

nilsnolde commented on conanfile.txt:
```diff
@@ -0,0 +1,6 @@
+[requires]
+openssl/1.1.1m
```
lastest is 3.0.1 but didn't try (too new, kinda pointless)

nilsnolde commented on conanfile.txt:
```diff
@@ -0,0 +1,6 @@
+openssl/1.1.1m
+asio/1.21.0
+
+[generators]
+json
```
this will instruct conan to write the build information of the 2 libraries to a json which we can then parse in the setup.py

nilsnolde commented on setup.py:
```diff
@@ -1,34 +1,70 @@
     extra_link_args.insert(0, "-L/usr/local/opt/openssl@1.1/lib")
+elif platform.system() == "Windows":
+    extra_compile_args = [
+        "-DNOGDI",
+        "-DNOMINMAX",
+        "-DWIN32_LEAN_AND_MEAN",
+        "-DASIO_STANDALONE"
```
I'm not sure anymore why `-DNOGDI`, `DNOMINMAX` was necessary though, there's some win trouble with `std::max` (also a problem with upstream), `DWIN32_LEAN_AND_MEAN` shaves off a few unneeded headers,

jonathf replied:
I have no idea. Let it be if it works.

nilsnolde commented on setup.py:
```diff
@@ -1,34 +1,70 @@
+        "-DWIN32_LEAN_AND_MEAN",
+        "-DASIO_STANDALONE"
+    ]
+    extra_link_args = []
```
reset link_args

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -2,26 +2,34 @@ name: Wheels
+      - name: Build SDist
+        run: pipx run build --sdist
 
-    - uses: actions/upload-artifact@v2
-      with:
-        path: dist/*.tar.gz
+      - name: Check metadata
+        run: pipx run twine check dist/*
 
+      - uses: actions/upload-artifact@v2
+        with:
+          path: dist/*.tar.gz
```
messed up diff.. nothing changed actually other than more consistent whitespace

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -32,16 +40,40 @@ jobs:
       with:
         submodules: recursive
 
+    - name: Cache Conan dependencies
+      id: cache-conan
+      uses: actions/cache@v2
+      if: matrix.os == 'windows-latest'
+      with:
+        path: $USERPROFILE/.conan
+        key: conan-${{ hashFiles('conanfile.txt') }}
+
+    - name: Set up Windows Python ${{ matrix.python_version }}
+      uses: actions/setup-python@v2
+      if: matrix.os == 'windows-latest'
+      with:
+        python-version: ${{ matrix.python_version }}
+
+    - name: Install conan for Windows
+      if: matrix.os == 'windows-latest' && steps.cache-conan.outputs.cache-hit != 'true'
+      run: |
+        pip install pip --upgrade
+        pip install conan
+        conan profile new default --detect
+        conan profile update "settings.compiler=Visual Studio" default
+        conan profile update "settings.compiler.version=16" default
+        conan install --build=openssl --install-folder conan_build .
+
```
I decided to take this here instead of `CIBW_BUILD_BEFORE_ALL` or so, since we want control when we attempt to install the conan dependencies (only if there was a cache miss). I didn't see a really good way to pass that on to cibuildwheel so it wouldn't execute a conan install conditionally. maybe env vars or so, but anyways, I think it's still fairly transparent. not sure yet if there's some linking issues when this is actually installed via `pip` on a target machine. there's no `.dll`s in the wheels, but I guess since the build links statically, there's simply no dependencies a windows build needs. we'll see the latest once the issue with numpy is resolved.

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -32,16 +40,42 @@ jobs:
       with:
         submodules: recursive
 
+    - name: Cache Conan dependencies
+      id: cache-conan
+      uses: actions/cache@v2
+      if: matrix.os == 'windows-latest'
+      with:
+        path:
+          $USERPROFILE/.conan
+          conan-build
+        key: conan-${{ hashFiles('conanfile.txt') }}-v1.1
+
+    - name: Set up Windows Python ${{ matrix.python_version }}
+      uses: actions/setup-python@v2
+      if: matrix.os == 'windows-latest'
+      with:
+        python-version: ${{ matrix.python_version }}
+
+    - name: Install conan for Windows
+      if: matrix.os == 'windows-latest' && steps.cache-conan.outputs.cache-hit != 'true'
+      run: |
+        pip install pip --upgrade
+        pip install conan
+        conan profile new default --detect
+        conan profile update "settings.compiler=Visual Studio" default
+        conan profile update "settings.compiler.version=16" default
+        conan install --build=openssl --install-folder conan_build .
```
github discarded it so again:

I decided to take this here instead of `CIBW_BUILD_BEFORE_ALL` or so, since we want control when we attempt to install the conan dependencies (only if there was a cache miss). I didn't see a really good way to pass that on to cibuildwheel so it wouldn't execute a conan install conditionally. maybe env vars or so, but anyways, I think it's still fairly transparent. not sure yet if there's some linking issues when this is actually installed via `pip` on a target machine. there's no `.dll`s in the wheels, but I guess since the build links statically, there's simply no dependencies a windows build needs. we'll see the latest once the issue with numpy is resolved.

nilsnolde commented:
I'll see tmrw if the cache for conan worked as well this time

jonathf commented on setup.py:
```diff
@@ -1,34 +1,70 @@
+    with open(conanfile[0]) as f:
+        conan_deps = json.load(f)['dependencies']
+    for dep in conan_deps:
+        include = dep['include_paths']
+        libs = dep['libs']
+        system_libs = dep['system_libs']
+        libdirs = dep['lib_paths']
+
+        include_dirs.extend(include)
+        if libs:
+            libraries.extend(libs)
+        if system_libs:
+            libraries.extend(system_libs)
+        if libdirs:
+            library_dirs.extend(libdirs)
```
Make this a bit more compact with:
```python
for dep in conan_deps:
    include_dirs.extend(dep["include_paths"])
    libraries.extend(dep["libs"])
    libraries.extend(dep["system_libs"])
    library_dirs.extend(dep["lib_paths"])
```
Extending with empty lists is fine.

jonathf reviewed (changes requested):
Amazing job, Nils!

A little bit of cleanup and this should be ready to merge.

jonathf commented:
> that should be it now I hope. I'll comment more in the PR. just one thing I don't quite get yet: how does cibuildwheel know what's in `[build-system].requires` of the `pyproject.toml`? is there some magic going on inside cibuildwheel which uses that value? locally I had to install pybind etc in my virtual env (I have a `build-requirements.txt` for pip I thought I'd contribute). but then CI obviously works, just don't understand how;)

PEP 517 is a python specification for placing it there. Most installers like `pip install`, `python -m build`, `poetry install`, etc. recognizes it.

jonathf commented:
BTW, I think you could get `python -m build` (which is what cibuildwheel uses) to work by adding `conanbuildinfo.json` to MANFEST.in.

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -2,26 +2,34 @@ name: Wheels
 
 on:
   push:
+    branches:
+      - 'main'
```
also depends a little how you usually manage your repos I guess. for me, I usually almost exclusively squash commits from PRs into main/master (other than hot fixes), then I really want to have a full build.. and then I limit the PR CI builds (e.g. have a 2nd yaml only building ubuntu 3.10?).

nilsnolde commented on README.md:
````diff
@@ -19,14 +19,30 @@ VROOM can also solve any mix of the above problem types.
 pip install pyvroom
 ```
 
+## Building
+
+Building the source distributions on another OS requires:
+- the `./build-requirements.txt` Python dependencies
+- `asio` headers installed
+- `openssl` & `crypto` libraries & headers installed
+
+Optionally the dependencies can be installed with [`conan`](https://github.com/conan-io/conan):
+```shell script
+conan install --build=openssl --install-folder conan_build .
+
+# note, on Windows you might have to execute the following before
+conan profile new default --detect
+conan profile update "settings.compiler=Visual Studio" default
+conan profile update "settings.compiler.version=16" default
+```
````
actually we can take that out of the readme as well, it's not really our responsibility, people should know how to wield conan if they choose to use it. or we'll eventually have a CONTRIBUTING.md detailing build setups.

while a bat file makes sense for CI, I think a user needs to know which msvc version is appropriate for them. is it ok for you to re-visit this in a separate PR, making an issue to remind ourselves, and I'll remove from the readme for now?

jonathf replied:
Yeah, that makes sinse.

nilsnolde replied:
included a bat file for CI and removed most of the section from the readme

nilsnolde commented on setup.py:
```diff
@@ -1,34 +1,70 @@
+library_dirs = []
+
+# try conan dependency resolution
+conanfile = tuple(Path().rglob('conanbuildinfo.json'))
+if conanfile:
+    print("INFO: Using conan to resolve dependencies.")
+    with open(conanfile[0]) as f:
+        conan_deps = json.load(f)['dependencies']
```
yeah right, should've left more comments.. 

- conanbuildinfo.json is a file that'll only be generate once you run `conan install` and the contents will somewhat depend on the build system (at least platform dependent). by default conan will put that file (and other aux files) into the `cwd` (which should be the repo or a subdirectory per convention). I for one usually use `conan install --install-folder conan_build` (also our CI uses that), that's why I `rglob`. so we need this file to be looked up dynamically, can't put in manifest.in either. 

rest makes perfect sense.

nilsnolde commented on setup.py:
```diff
@@ -1,34 +1,70 @@
+    print('WARN: Conan not installed and/or no conan build detected. Assuming dependencies are installed.')
 
 if platform.system() == "Darwin":
     # Homebrew puts include folders in weird places.
     include_dirs.append("/usr/local/opt/openssl@1.1/include")
     extra_link_args.insert(0, "-L/usr/local/opt/openssl@1.1/lib")
```
actually, using conan on mac for ci would remove these lines as well. which are not very robust I think. brew is pretty updated always I found, some users might actually have 3.x soon? anyways, more of a works-now-why-change thing. but maybe good to keep in mind in worst case

nilsnolde commented on conan_install.bat:
```diff
@@ -0,0 +1,4 @@
+conan profile new default --detect
+conan profile update "settings.compiler=Visual Studio" default
+conan profile update "settings.compiler.version=16" default
+conan install --build=openssl --install-folder conan_build .
```
didn't do bat files in a long time, but I'd guess this is enough.

nilsnolde commented:
addressed your review, thanks, was really helpful!

win CI is not caching right yet. might need to do a round or two more.. generally no stress with this (from my side anyways). I seem super active right now, but just trying to get it out of my head;)

nilsnolde commented:
jeez.. that ruined my day really, as much as I hate letting things go, this is one of these times.. for now anyways.. 

env vars in github actions are TERRIBLE.. I really tried 3-4 different ways of doing it, none worked (also doesn't help that it's not really stable yet and keeps changing).. this is the official guide: https://docs.github.com/en/actions/learn-github-actions/workflow-commands-for-github-actions#environment-files. no idea how to work these environment files and the section below that (>> $GITHUB_ENV) didn't work either. 

if you have a good idea, most appreciated.. the problem is that:
- we need the path to cache for conan's home (that's where the libs live). it can be hard-coded (it's at `C:\Users\runneradmin\.conan`), but not even that works..
- cache the folder conan installs the aux files to (`./conan_build`) as per `conan_install.bat`

jonathf commented:
Sorry, I have very little expertice on Windows. My guess is likely going to be worse than yours.

jonathf commented:
Sorry, wrong button...

nilsnolde commented:
no worries, next time I'll have more patience. problem is github actions, not so much windows. it just won't cache the path I tell it to. win desktop is one frustrating experience, but windows CI takes the cake!

nilsnolde commented:
oh boy... had a long battle with win ci in our fork.. seems like I slayed the bastard finally! 

problems were many.. the cache action somehow uses gnu tar.exe which seems to have problems with windows path separators. and github actions env vars on windows have to be prefixed with `$env:`, e.g. `$env:GITHUB_WORKSPACE`. then there's 3 different shells to use, each worse than the other and and and..

nilsnolde commented on .github/workflows/wheels.yml:
```diff
@@ -32,16 +40,43 @@ jobs:
+      run: |
+        pip install pip --upgrade
+        pip install conan
+        conan profile new default --detect
+        conan profile update "settings.compiler=Visual Studio" default
+        conan profile update "settings.compiler.version=16" default
+        conan config set "storage.path=$env:GITHUB_WORKSPACE/conan_data"
+        conan install --build=openssl --install-folder conan_build .
```
I took this back instead of the conan_install.bat. we need the `$env:GITHUB_WORKSPACE`. after this worked in your repo with all jobs active I'll try again with an env var in the .bat file. it _should_ work, but honestly if it doesn't on the first try, I'll leave this as it is now..

nilsnolde replied:
ran through, but I think we can leave as it is in any case and accept that windows is just bääh;)

jonathf reviewed (approved):
Great work @nilsnolde! I know how difficult win build can be.
